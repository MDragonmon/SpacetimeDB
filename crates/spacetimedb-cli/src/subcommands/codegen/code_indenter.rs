use std::fmt;
use std::ops::{Deref, DerefMut};

pub struct CodeIndenter<W: fmt::Write> {
    writer: W,
    level: u32,
    needs_indenting: bool,
}
impl<W: fmt::Write> CodeIndenter<W> {
    pub fn new(writer: W) -> Self {
        CodeIndenter {
            writer,
            level: 0,
            needs_indenting: true,
        }
    }
    // pub fn get_ref(&self) -> &W {
    //     &self.writer
    // }
    // pub fn get_mut(&mut self) -> &mut W {
    //     &mut self.writer
    // }
    pub fn into_inner(self) -> W {
        self.writer
    }
    pub fn indent(&mut self, n: u32) {
        self.level = self.level.saturating_add(n);
    }
    pub fn dedent(&mut self, n: u32) {
        self.level = self.level.saturating_sub(n);
    }
    pub fn indented(&mut self, n: u32) -> IndentScope<'_, W> {
        self.indent(n);
        IndentScope { fmt: self }
    }
    fn write_indent(&mut self) -> fmt::Result {
        for _ in 0..self.level {
            self.writer.write_str(super::INDENT)?;
        }
        Ok(())
    }
}
impl<W: fmt::Write> fmt::Write for CodeIndenter<W> {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        for (i, line) in s.split_inclusive('\n').enumerate() {
            let write_indent = i != 0 || std::mem::take(&mut self.needs_indenting);
            // skip the indent if it's an empty line
            if write_indent && line != "\n" {
                self.write_indent()?;
            }
            self.writer.write_str(line)?;
        }
        self.needs_indenting = s.ends_with('\n');
        Ok(())
    }
}
pub struct IndentScope<'a, W: fmt::Write> {
    fmt: &'a mut CodeIndenter<W>,
}
impl<W: fmt::Write> Drop for IndentScope<'_, W> {
    fn drop(&mut self) {
        self.fmt.dedent(1);
    }
}
impl<T: fmt::Write> Deref for IndentScope<'_, T> {
    type Target = CodeIndenter<T>;
    fn deref(&self) -> &Self::Target {
        self.fmt
    }
}
impl<T: fmt::Write> DerefMut for IndentScope<'_, T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.fmt
    }
}
