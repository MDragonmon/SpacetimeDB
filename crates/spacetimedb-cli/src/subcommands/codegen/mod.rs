use std::fs;
use std::path::Path;

use crate::util;
use clap::Arg;
use convert_case::{Case, Casing};
use spacetimedb_lib::type_def::{ReducerDef, TableDef};
use spacetimedb_lib::TupleDef;
use wasmtime::{ExternType, Trap, TypedFunc};

mod code_indenter;
pub mod csharp;

const INDENT: &str = "\t";

pub fn cli() -> clap::Command<'static> {
    clap::Command::new("generate")
        .about("Generate client files for a spacetime module.")
        .arg(
            Arg::new("wasm_file")
                .takes_value(true)
                .required(false)
                .long("wasm-file")
                .short('w')
                .conflicts_with("projec_path"),
        )
        .arg(
            Arg::new("project_path")
                .takes_value(true)
                .required(false)
                .long("project-path")
                .short('p')
                .default_value(".")
                .conflicts_with("wasm_file"),
        )
        .arg(
            Arg::new("out_dir")
                .takes_value(true)
                .required(true)
                .long("out-dir")
                .short('o'),
        )
        .arg(
            Arg::new("lang")
                .takes_value(true)
                .required(true)
                .long("lang")
                .short('l')
                .possible_values(["csharp", "cs", "c#"]),
        )
        .after_help("Run `spacetime help publish` for more detailed information.")
}

pub fn exec(args: &clap::ArgMatches) -> anyhow::Result<()> {
    let project_path = args.value_of("project_path").unwrap();
    let wasm_file_path = Path::new(project_path);
    let wasm_file_path = util::find_wasm_file(wasm_file_path)?;
    let wasm_file = match args.value_of("wasm_file") {
        None => Path::new(wasm_file_path.to_str().unwrap()),
        Some(path) => Path::new(path),
    };

    let out_dir = Path::new(args.value_of("out_dir").unwrap());
    let lang = args.value_of("lang").unwrap();

    if !out_dir.exists() {
        return Err(anyhow::anyhow!(
            "Output directory '{}' does not exist. Please create the directory and rerun this command.",
            out_dir.to_str().unwrap()
        ));
    }

    for (fname, code) in generate(wasm_file, lang)? {
        fs::write(out_dir.join(fname), code)?;
    }
    Ok(())
}

pub fn generate(wasm_file: &Path, lang: &str) -> anyhow::Result<impl Iterator<Item = (String, String)>> {
    match lang {
        "csharp" | "cs" | "c#" => {
            let descriptions = extract_descriptions(&wasm_file)?;
            Ok(descriptions.into_iter().map(|(name, desc)| match desc {
                Description::Table(table) => {
                    let code = csharp::autogen_csharp_table(&name, &table);
                    (name + ".cs", code)
                }
                Description::Tuple(tup) => {
                    let code = csharp::autogen_csharp_tuple(&name, &tup);
                    (name + ".cs", code)
                }
                Description::Reducer(reducer) => {
                    let code = csharp::autogen_csharp_reducer(&reducer);
                    let pascalcase = name.to_case(Case::Pascal);
                    (pascalcase + "Reducer.cs", code)
                }
            }))
        }
        &_ => Err(anyhow::anyhow!(format!("Unsupported langauge: {}", lang))),
    }
}

enum Description {
    Table(TableDef),
    Tuple(TupleDef),
    Reducer(ReducerDef),
}

fn extract_descriptions(wasm_file: &Path) -> anyhow::Result<Vec<(String, Description)>> {
    let engine = wasmtime::Engine::default();
    let module = wasmtime::Module::from_file(&engine, wasm_file)?;
    let mut store = wasmtime::Store::new(&engine, ());
    let mut linker = wasmtime::Linker::new(&engine);
    for imp in module.imports() {
        if let ExternType::Func(func_type) = imp.ty() {
            linker
                .func_new(imp.module(), imp.name(), func_type, |_, _, _| {
                    Err(Trap::new("don't call me!!"))
                })
                .unwrap();
        }
    }
    let instance = linker.instantiate(&mut store, &module)?;
    let memory = instance.get_memory(&mut store, "memory").unwrap();
    // let alloc: TypedFunc<(u32,), (u32,)> = instance.get_func(&mut store, "alloc").unwrap().typed(&store).unwrap();
    let dealloc: TypedFunc<(u32, u32), ()> = instance.get_func(&mut store, "dealloc").unwrap().typed(&store).unwrap();
    enum DescrType {
        Table,
        Tuple,
        Reducer,
    }
    let describes = instance
        .exports(&mut store)
        .filter_map(|exp| {
            let sym = exp.name();
            None.or_else(|| sym.strip_prefix("__describe_table__").map(|n| (DescrType::Table, n)))
                .or_else(|| sym.strip_prefix("__describe_tuple__").map(|n| (DescrType::Tuple, n)))
                .or_else(|| {
                    sym.strip_prefix("__describe_reducer__")
                        .map(|n| (DescrType::Reducer, n))
                })
                .map(|(ty, name)| (ty, name.to_owned(), exp.into_func().unwrap()))
        })
        .collect::<Vec<_>>();
    let mut descriptions = Vec::with_capacity(describes.len());
    for (ty, name, describe) in describes {
        let describe: TypedFunc<(), (u64,)> = describe.typed(&store).unwrap();
        let (val,) = describe.call(&mut store, ()).unwrap();
        let offset = (val >> 32) as u32;
        let len = (val & 0xFFFF_FFFF) as u32;
        let slice = &memory.data(&store)[offset as usize..][..len as usize];
        let descr = match ty {
            DescrType::Table => Description::Table(TableDef::decode(&mut &slice[..])?),
            DescrType::Tuple => Description::Tuple(TupleDef::decode(&mut &slice[..])?),
            DescrType::Reducer => Description::Reducer(ReducerDef::decode(&mut &slice[..])?),
        };
        dealloc.call(&mut store, (offset, len)).unwrap();
        descriptions.push((name, descr));
    }
    Ok(descriptions)
}
