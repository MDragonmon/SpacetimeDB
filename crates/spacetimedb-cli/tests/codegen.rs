use std::collections::HashMap;
use std::path::Path;

#[test]
fn test_codegen_output() {
    let path = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../target/wasm32-unknown-unknown/release/rust_wasm_test.wasm"
    ));
    if !path.exists() {
        eprintln!("rust_wasm_test isn't built, skipping");
    }
    use spacetimedb_cli::codegen;
    let outfiles: HashMap<_, _> = codegen::gen_bindings(path, codegen::Language::Csharp)
        .unwrap()
        .collect();
    insta::with_settings!({ sort_maps => true }, {
        insta::assert_toml_snapshot!(outfiles);
    });
}
