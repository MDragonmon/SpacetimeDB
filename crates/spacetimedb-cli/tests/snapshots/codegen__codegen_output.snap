---
source: crates/spacetimedb-cli/tests/codegen.rs
expression: outfiles
---
"TestA.cs" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

using System;

namespace SpacetimeDB
{
	public partial class TestA : IDatabaseTable
	{
		[Newtonsoft.Json.JsonProperty("x")]
		public uint x;
		[Newtonsoft.Json.JsonProperty("y")]
		public uint y;
		[Newtonsoft.Json.JsonProperty("z")]
		public string z;

		public static event Action<TestA> OnInsert;
		public static event Action<TestA, TestA> OnUpdate;
		public static event Action<TestA> OnDelete;
		public static event Action<NetworkManager.TableOp, TestA, TestA> OnRowUpdate;

		public static TypeDef GetTypeDef()
		{
			return TypeDef.Tuple(new ElementDef[]
			{
				new SpacetimeDB.ElementDef(0, SpacetimeDB.TypeDef.BuiltInType(SpacetimeDB.TypeDef.Def.U32)),
				new SpacetimeDB.ElementDef(1, SpacetimeDB.TypeDef.BuiltInType(SpacetimeDB.TypeDef.Def.U32)),
				new SpacetimeDB.ElementDef(2, SpacetimeDB.TypeDef.BuiltInType(SpacetimeDB.TypeDef.Def.String)),
			});
		}

		public static explicit operator TestA(TypeValue value)
		{
			var tupleValue = value.GetValue(TypeDef.Def.Tuple) as TypeValue[];
			if (tupleValue == null)
			{
				throw new System.InvalidOperationException($"Invalid value (must be Tuple): {value.TypeDef.Type}");
			}

			return new TestA
			{
				x = (uint)tupleValue[0].GetValue(TypeDef.Def.U32),
				y = (uint)tupleValue[1].GetValue(TypeDef.Def.U32),
				z = (string)tupleValue[2].GetValue(TypeDef.Def.String),
			};
		}
		public static System.Collections.Generic.IEnumerable<TestA> Iter()
		{
			foreach(var entry in NetworkManager.clientDB.GetEntries("TestA"))
			{
				var tupleArr = entry.GetValue(TypeDef.Def.Tuple) as TypeValue[];
				if (tupleArr == null) continue;
				yield return (TestA)entry;
			}
		}
		public static System.Collections.Generic.IEnumerable<TestA> FilterByX(uint value)
		{
			foreach(var entry in NetworkManager.clientDB.GetEntries("TestA"))
			{
				var tupleArr = entry.GetValue(TypeDef.Def.Tuple) as TypeValue[];
				if (tupleArr == null) continue;
				var compareValue = (uint)tupleArr[0].GetValue(TypeDef.Def.U32);
				if (compareValue == value)
					yield return (TestA)entry;
			}
		}

		public static System.Collections.Generic.IEnumerable<TestA> FilterByY(uint value)
		{
			foreach(var entry in NetworkManager.clientDB.GetEntries("TestA"))
			{
				var tupleArr = entry.GetValue(TypeDef.Def.Tuple) as TypeValue[];
				if (tupleArr == null) continue;
				var compareValue = (uint)tupleArr[1].GetValue(TypeDef.Def.U32);
				if (compareValue == value)
					yield return (TestA)entry;
			}
		}

		public static System.Collections.Generic.IEnumerable<TestA> FilterByZ(string value)
		{
			foreach(var entry in NetworkManager.clientDB.GetEntries("TestA"))
			{
				var tupleArr = entry.GetValue(TypeDef.Def.Tuple) as TypeValue[];
				if (tupleArr == null) continue;
				var compareValue = (string)tupleArr[2].GetValue(TypeDef.Def.String);
				if (compareValue == value)
					yield return (TestA)entry;
			}
		}

		public static void OnInsertEvent(object newValue)
		{
			if(OnInsert != null)
			{
				OnInsert?.Invoke((TestA)newValue);
			}
		}

		public static void OnUpdateEvent(object oldValue, object newValue)
		{
			if(OnUpdate != null)
			{
				OnUpdate?.Invoke((TestA)oldValue,(TestA)newValue);
			}
		}

		public static void OnDeleteEvent(object oldValue)
		{
			if(OnDelete != null)
			{
				OnDelete?.Invoke((TestA)oldValue);
			}
		}

		public static void OnRowUpdateEvent(NetworkManager.TableOp op, object oldValue, object newValue)
		{
			if(OnRowUpdate != null)
			{
				OnRowUpdate?.Invoke(op, (TestA)oldValue,(TestA)newValue);
			}
		}
	}
}
'''
"TestB.cs" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

using System;

namespace SpacetimeDB
{
	public partial class TestB : IDatabaseTable
	{
		[Newtonsoft.Json.JsonProperty("foo")]
		public string foo;

		public static event Action<TestB> OnInsert;
		public static event Action<TestB, TestB> OnUpdate;
		public static event Action<TestB> OnDelete;
		public static event Action<NetworkManager.TableOp, TestB, TestB> OnRowUpdate;

		public static TypeDef GetTypeDef()
		{
			return TypeDef.Tuple(new ElementDef[]
			{
				new SpacetimeDB.ElementDef(0, SpacetimeDB.TypeDef.BuiltInType(SpacetimeDB.TypeDef.Def.String)),
			});
		}

		public static explicit operator TestB(TypeValue value)
		{
			var tupleValue = value.GetValue(TypeDef.Def.Tuple) as TypeValue[];
			if (tupleValue == null)
			{
				throw new System.InvalidOperationException($"Invalid value (must be Tuple): {value.TypeDef.Type}");
			}

			return new TestB
			{
				foo = (string)tupleValue[0].GetValue(TypeDef.Def.String),
			};
		}
		public static void OnInsertEvent(object newValue)
		{
			if(OnInsert != null)
			{
				OnInsert?.Invoke((TestB)newValue);
			}
		}

		public static void OnUpdateEvent(object oldValue, object newValue)
		{
			if(OnUpdate != null)
			{
				OnUpdate?.Invoke((TestB)oldValue,(TestB)newValue);
			}
		}

		public static void OnDeleteEvent(object oldValue)
		{
			if(OnDelete != null)
			{
				OnDelete?.Invoke((TestB)oldValue);
			}
		}

		public static void OnRowUpdateEvent(NetworkManager.TableOp op, object oldValue, object newValue)
		{
			if(OnRowUpdate != null)
			{
				OnRowUpdate?.Invoke(op, (TestB)oldValue,(TestB)newValue);
			}
		}
	}
}
'''
"TestReducer.cs" = '''
// THIS FILE IS AUTOMATICALLY GENERATED BY SPACETIMEDB. EDITS TO THIS FILE
// WILL NOT BE SAVED. MODIFY TABLES IN RUST INSTEAD.

using System;
using ClientApi;
using Newtonsoft.Json.Linq;

namespace SpacetimeDB
{
	public static partial class Reducer
	{
		public static event Action<Event.Types.Status, Hash, TestA, TestB> OnTestEvent;

		public static void Test(TestA arg, TestB arg2)
		{
			NetworkManager.instance.InternalCallReducer(new NetworkManager.Message
			{
				fn = "test",
				args = new object[] { arg, arg2 },
			});
		}

		[ReducerEvent(FunctionName = "test")]
		public static void OnTest(Event dbEvent)
		{
			if(OnTestEvent != null)
			{
				var jsonString = dbEvent.FunctionCall.ArgBytes.ToStringUtf8();
				var args = Newtonsoft.Json.JsonConvert.DeserializeObject<JArray>(jsonString);
				if(args.Count >= 2)
				{
					OnTestEvent(dbEvent.Status, Hash.From(dbEvent.CallerIdentity.ToByteArray()), args[0].ToObject<TestA>(), args[1].ToObject<TestB>());
				}
			}
		}
	}
}
'''
